imap/monitor.go

```diff
@@ -0,0 +1,212 @@
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			dbusers, err := models.GetUsers() //Slice of all user ids. Each user gets their own IMAP monitor routine.
+			if err != nil {
+				log.Error(err)
+				break
+			}
+			for _, dbuser := range dbusers {
+				if _, ok := usermap[dbuser.Id]; !ok { // If we don't currently have a running Go routine for this user, start one.
+					log.Info("Starting new IMAP monitor for user ", dbuser.Username)
+					usermap[dbuser.Id] = 1
+					go monitorIMAP(dbuser.Id, ctx, im.reportURL)
+				}
+			}
+			time.Sleep(10 * time.Second) // Every ten seconds we check if a new user has been created
+		}
+	}
+}
+
+// monitorIMAP will continuously login to the IMAP settings associated to the supplied user id (if the user account has IMAP settings, and they're enabled.)
+// It also verifies the user account exists, and returns if not (for the case of a user being deleted).
+func monitorIMAP(uid int64, ctx context.Context, reportURL string) {
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			// 1. Check if user exists, if not, return.
+			_, err := models.GetUser(uid)
+			if err != nil { // Not sure if there's a better way to determine user existence via id.
+				log.Info("User ", uid, " seems to have been deleted. Stopping IMAP monitor for this user.")
+				return
+			}
+			// 2. Check if user has IMAP settings.
+			imapSettings, err := models.GetIMAP(uid)
+			if err != nil {
+				log.Error(err)
+				break
+			}
+			if len(imapSettings) > 0 {
+				im := imapSettings[0]
+				// 3. Check if IMAP is enabled
+				if im.Enabled {
+					log.Debug("Checking IMAP for user ", uid, ": ", im.Username, "@", im.Host)
+					checkForNewEmails(im, reportURL)
+					time.Sleep((time.Duration(im.IMAPFreq) - 10) * time.Second) // Subtract 10 to compensate for the default sleep of 10 at the bottom
+				}
+			}
+		}
+		time.Sleep(10 * time.Second)
+	}
+}
+
+// NewMonitor returns a new instance of imap.Monitor
+func NewMonitor(config *config.Config) *Monitor {
+
+	reportURL := "http://" + config.PhishConf.ListenURL
+	if config.PhishConf.UseTLS {
+		reportURL = "https://" + config.PhishConf.ListenURL
+	}
+
+	im := &Monitor{
+		reportURL: reportURL,
+	}
+	return im
+}
+
+// Start launches the IMAP campaign monitor
+func (im *Monitor) Start() error {
+	log.Info("Starting IMAP monitor manager")
+	ctx, cancel := context.WithCancel(context.Background()) // ctx is the derivedContext
+	im.cancel = cancel
+	go im.start(ctx)
+	return nil
+}
+
+// Shutdown attempts to gracefully shutdown the IMAP monitor.
+func (im *Monitor) Shutdown() error {
+	log.Info("Shutting down IMAP monitor manager")
+	im.cancel()
+	return nil
+}
+
+// checkForNewEmails logs into an IMAP account and checks unread emails
+//  for the rid campaign identifier.
+func checkForNewEmails(im models.IMAP, reportURL string) {
+
+	mailSettings := MailboxInfo{
+		Host:   im.Host,
```

Comment:
There's a bug right now where the `Port` in the `models.IMAP` object doesn't appear to be used in the `MailboxInfo` object. This appears to be the case in both the validation and general IMAP monitoring.

This causes `newIMAPClient` to call `DialTLS` or `Dial` with just the hostname, causing `github.com/mxk/go-imap/imap` to choose the default port (993 for TLS, 143 without).

This appears to only affect the IMAP polling. You do add the port in the IMAP validation.